File: Polars.CSharp/Expr/Lit.cs
using Polars.NET.Core;

namespace Polars.CSharp;

/// <summary>
/// A Polars Expr
/// </summary>
public partial class Expr : IDisposable
{
    internal static Expr MakeLit(object val)
    {
        if (val is Expr e) return e;

        if (val is Selector s) return s.ToExpr();

        return val switch
        {
            // --- Integer ---
            int i => new Expr(PolarsWrapper.Lit(i)),
            uint ui => new Expr(PolarsWrapper.Lit(ui)),
            long l => new Expr(PolarsWrapper.Lit(l)),
            ulong ul => new Expr(PolarsWrapper.Lit(ul)),
            short sh => new Expr(PolarsWrapper.Lit(sh)),
            ushort ush => new Expr(PolarsWrapper.Lit(ush)),
            byte by => new Expr(PolarsWrapper.Lit(by)),
            sbyte sb => new Expr(PolarsWrapper.Lit(sb)),
            Int128 i128 => new Expr(PolarsWrapper.Lit(i128)),

            // --- Float ---
            double d => new Expr(PolarsWrapper.Lit(d)),
            float f => new Expr(PolarsWrapper.Lit(f)),
            Half h => new Expr(PolarsWrapper.Lit(h)),

            // --- String and Boolean ---
            string str => new Expr(PolarsWrapper.Lit(str)),
            bool b => new Expr(PolarsWrapper.Lit(b)),

            // --- Time ---
            DateTime dt => new Expr(PolarsWrapper.Lit(dt)),
            DateTimeOffset dtos => new Expr(PolarsWrapper.Lit(dtos)),
            DateOnly date => new Expr(PolarsWrapper.Lit(date)),
            TimeOnly time => new Expr(PolarsWrapper.Lit(time)),
            TimeSpan ts => new Expr(PolarsWrapper.Lit(ts)),

            // --- Decimal ---
            decimal dec => Polars.Lit(dec),

            // --- Null ---
            null => new Expr(PolarsWrapper.LitNull()),

            // --- Exception ---
            _ => throw new NotSupportedException($"Unsupported literal type: {val.GetType().Name}")
        };
    }

}

File: Polars.FSharp/Lit.fs
namespace Polars.FSharp

open System.ComponentModel
open Polars.NET.Core
open System

[<EditorBrowsable(EditorBrowsableState.Never)>]
type LitMechanism = LitMechanism with
    static member ($) (LitMechanism, v: int8) = new Expr(PolarsWrapper.Lit v)
    static member ($) (LitMechanism, v: uint8) = new Expr(PolarsWrapper.Lit v)
    static member ($) (LitMechanism, v: int16) = new Expr(PolarsWrapper.Lit v)
    static member ($) (LitMechanism, v: uint16) = new Expr(PolarsWrapper.Lit v)
    static member ($) (LitMechanism, v: int) = new Expr(PolarsWrapper.Lit v)
    static member ($) (LitMechanism, v: uint) = new Expr(PolarsWrapper.Lit v)
    static member ($) (LitMechanism, v: int64) = new Expr(PolarsWrapper.Lit v)
    static member ($) (LitMechanism, v: uint64) = new Expr(PolarsWrapper.Lit v)
    static member ($) (LitMechanism, v: Int128) = new Expr(PolarsWrapper.Lit v)
    static member ($) (LitMechanism, v: string) = new Expr(PolarsWrapper.Lit v)
    static member ($) (LitMechanism, v: Half) = new Expr(PolarsWrapper.Lit v)
    static member ($) (LitMechanism, v: float32) = new Expr(PolarsWrapper.Lit v)
    static member ($) (LitMechanism, v: double) = new Expr(PolarsWrapper.Lit v)
    static member ($) (LitMechanism, v: DateTime) = new Expr(PolarsWrapper.Lit v)
    static member ($) (LitMechanism, v: DateOnly) = new Expr(PolarsWrapper.Lit v)
    static member ($) (LitMechanism, v: TimeOnly) = new Expr(PolarsWrapper.Lit v)
    static member ($) (LitMechanism, v: DateTimeOffset) = new Expr(PolarsWrapper.Lit v)
    static member ($) (LitMechanism, v: TimeSpan) = new Expr(PolarsWrapper.Lit v)
    static member ($) (LitMechanism, v: decimal) = new Expr(PolarsWrapper.Lit v)
    static member ($) (LitMechanism, v: bool) = new Expr(PolarsWrapper.Lit v)
    // --- List ---
    static member ($) (LitMechanism, v: int list)      = new Expr(PolarsWrapper.Lit(Series.create("", v).Handle))
    static member ($) (LitMechanism, v: int64 list)    = new Expr(PolarsWrapper.Lit(Series.create("", v).Handle))
    static member ($) (LitMechanism, v: Half list)     = new Expr(PolarsWrapper.Lit(Series.create("", v).Handle))
    static member ($) (LitMechanism, v: float list)    = new Expr(PolarsWrapper.Lit(Series.create("", v).Handle)) // double
    static member ($) (LitMechanism, v: float32 list)  = new Expr(PolarsWrapper.Lit(Series.create("", v).Handle))
    static member ($) (LitMechanism, v: string list)   = new Expr(PolarsWrapper.Lit(Series.create("", v).Handle))
    static member ($) (LitMechanism, v: bool list)     = new Expr(PolarsWrapper.Lit(Series.create("", v).Handle))
    static member ($) (LitMechanism, v: DateTime list) = new Expr(PolarsWrapper.Lit(Series.create("", v).Handle))
    static member ($) (LitMechanism, v: DateOnly list) = new Expr(PolarsWrapper.Lit(Series.create("", v).Handle))

    // --- Array (High Performance) ---
    static member ($) (LitMechanism, v: int[])      = new Expr(PolarsWrapper.Lit(Series.create("", v).Handle))
    static member ($) (LitMechanism, v: Half[])    = new Expr(PolarsWrapper.Lit(Series.create("", v).Handle))
    static member ($) (LitMechanism, v: int64[])    = new Expr(PolarsWrapper.Lit(Series.create("", v).Handle))
    static member ($) (LitMechanism, v: float[])    = new Expr(PolarsWrapper.Lit(Series.create("", v).Handle))
    static member ($) (LitMechanism, v: float32[])  = new Expr(PolarsWrapper.Lit(Series.create("", v).Handle))
    static member ($) (LitMechanism, v: string[])   = new Expr(PolarsWrapper.Lit(Series.create("", v).Handle))
    static member ($) (LitMechanism, v: bool[])     = new Expr(PolarsWrapper.Lit(Series.create("", v).Handle))
    static member ($) (LitMechanism, v: DateTime[]) = new Expr(PolarsWrapper.Lit(Series.create("", v).Handle))

    // --- Nullable List (Option) ---
    static member ($) (LitMechanism, v: int option list)    = new Expr(PolarsWrapper.Lit(Series.create("", v).Handle))
    static member ($) (LitMechanism, v: float option list)  = new Expr(PolarsWrapper.Lit(Series.create("", v).Handle))
    static member ($) (LitMechanism, v: string option list) = new Expr(PolarsWrapper.Lit(Series.create("", v).Handle))
